ARROW-17453: [Go][C++][Parquet] Inconsistent Data with Repetition Levels

zeroshade opened the pull request:
Both the C++ and Go parquet implementations assumed that if the max repetition level was 0, that there were no bytes to be skipped when initializing decoders for `DataPageV2` but the Parquet files produced by Athena in this case had repetition bytes to be skipped before getting the definition level bytes. Since the byte wasn't skipped, the wrong values were decoded for Definition levels.

In the case of the Go implementation, it made additional assumptions that proved to be incorrect on top of the same bug.

This fixes both of them to properly respect the repetition level byte length reported in the DataPageV2 header.

emkornfield commented on cpp/src/parquet/column_reader_test.cc:
```diff
@@ -356,6 +356,34 @@ TEST_F(TestPrimitiveReader, TestReadValuesMissing) {
+  max_rep_level_ = 0;
+  NodePtr type = schema::Int32("a", Repetition::OPTIONAL);
+  const ColumnDescriptor descr(type, max_def_level_, max_rep_level_);
+  const std::array<uint8_t, 21> page_data{0x3,  0x3, 0x7, 0x80, 0x1, 0x4, 0x3,
```
it is possble to provide some explanation of where these bytes cape from?

zeroshade replied:
sure, in this case i just took them from the example parquet file in the JIRA card, the important bit is the first three bytes. If the first byte isn't skipped then RLE decoding yields [1, 1, 0, 0] instead of the correct [1, 1, 1, 0] for the definition levels. I'll add a comment.

emkornfield replied:
yeah documented the important piece would be useful.

emkornfield commented on cpp/src/parquet/column_reader_test.cc:
```diff
@@ -356,6 +356,34 @@ TEST_F(TestPrimitiveReader, TestReadValuesMissing) {
+  max_rep_level_ = 0;
+  NodePtr type = schema::Int32("a", Repetition::OPTIONAL);
+  const ColumnDescriptor descr(type, max_def_level_, max_rep_level_);
+  const std::array<uint8_t, 21> page_data{0x3,  0x3, 0x7, 0x80, 0x1, 0x4, 0x3,
```
I don't think there is a strong reason to use array here, as it requires setting the actual number of bytes, you could just use a vector.

zeroshade commented:
Added comment with extra clarification and changed `std::array` to `std::vector`.

Also, this is the same issue in https://issues.apache.org/jira/browse/ARROW-17452 so I resolved it as a duplicate.

zeroshade commented:
R failure is unrelated to this change.

@thisisnic @dragosmg could one of you or someone else look into it when you get a chance? I think it's a flaky workflow and spurious failure... but i could be wrong.

dragosmg commented:
> R failure is unrelated to this change.
> 
> @thisisnic @dragosmg could one of you or someone else look into it when you get a chance? I think it's a flaky workflow and spurious failure... but i could be wrong.

The error appeared after #13665 was merged, so definitely unrelated to this change (but also not spurious). It's present in some of my recent PRs too. See #13960.

zeroshade commented:
@dragosmg thanks for the confirmation. Once the last couple checks pass i'll merge this.
